corehq/apps/app_manager/models.py

```diff
@@ -2226,6 +2226,7 @@ class ModuleBase(IndexedSchema, ModuleMediaMixin, NavMenuItemMediaMixin, Comment
     auto_select_case = BooleanProperty(default=False)
     is_training_module = BooleanProperty(default=False)
     session_endpoint_id = StringProperty(exclude_if_none=True)  # See toggles.SESSION_ENDPOINTS
+    custom_assertions = SchemaListProperty(CustomAssertion)
```

Comment:
Reviewers: I'd love your take on whether to add this to `ModuleBase`, as I've done here, versus `Module`.  This seemed to make the most sense, as conceptually, assertions could logically apply to `ReportModule`s and `ShadowModule`s too, whether or not we make that possible in the UI

Reply:
What do you think are the odds a shadow module would want its own assertion, versus inheriting from its source module?

Reply:
Conceptually this seems like the right place to me though if we don't implement it for other module types it's a bit weird to have it in the model. If it's not hard to implement for all module types I would just do that (perhaps with the exception of ShadowModules - at least for now)

Reply:
I think it should be possible to do it for all module types - it looks like they use a shared template for the settings page (except for I think report modules).

> What do you think are the odds a shadow module would want its own assertion, versus inheriting from its source module?

I was initially thinking shadow module assertions would be added on top of any assertions in its source module, but now I'm not so sure.  If the assertions are present to encode assumptions used in the module's forms, then yes inheriting from the source module makes sense.  However if assertions are there to assist with say case list, detail, or search, then having their own assertions makes sense.

Where I'm landing now is that shadow modules should get their own assertions, and not inherit from the source module.  Maybe it's a cop-out, but it doesn't make any assumptions about the purpose about the assertions, and allows the most flexibility.  If source module assertions _should_ apply, app builders can always copy them over (or move them to the app level, if appropriate).